Adapter/Algolia.php

```diff
@@ -90,140 +96,200 @@ public function __construct(
     }
 
-    /**
-     * Get native query documents
-     *
-     * @param  RequestInterface $request
-     *
-     * @return array
-     */
-    public function getNativeQueryData($request)
+    private function nativeQuery(RequestInterface $request)
     {
-        $query  = $this->mapper->buildQuery($request);
+        $query = $this->mapper->buildQuery($request);
         $temporaryStorage = $this->temporaryStorageFactory->create();
         $table = $temporaryStorage->storeDocumentsFromSelect($query);
+
         $documents = $this->getDocuments($table);
 
-        return [
+        $aggregations = $this->aggregationBuilder->build($request, $table, $documents);
+        $response = [
             'documents' => $documents,
-            'table' => $table,
+            'aggregations' => $aggregations,
         ];
+
+        return $this->responseFactory->create($response);
+    }
+
+    private function getApiDocument($document)
+    {
+        return $this->documentFactory->create($document);
     }
 
     /**
-     * Checks if Algolia is properly configured and enabled
+     * Get search result from Algolia
      *
-     * @param  int     $storeId
+     * @param string $algoliaQuery
+     * @param int $storeId
      *
-     * @return bool
+     * @return array
      */
-    private function isAllowed($storeId)
+    private function getDocumentsFromAlgolia($algoliaQuery, $storeId)
     {
-        return
-            $this->config->getApplicationID($storeId)
-            && $this->config->getAPIKey($storeId)
-            && $this->config->isEnabledFrontEnd($storeId)
-            && $this->config->makeSeoRequest($storeId);
+        $contextParams = [];
+        $targetedIndex = null;
+        if ($this->isReplaceCategory($storeId)) {
+            $contextParams = $this->getContextParams($storeId);
+
+            if (!is_null($this->request->getParam('sortBy'))) {
+                $targetedIndex = $this->request->getParam('sortBy');
+            }
+        }
+
+        return $this->algoliaHelper->getSearchResult($algoliaQuery, $storeId, $contextParams, $targetedIndex);
     }
 
     /**
-     * Get aggregations
+     * Get the contextual params from the url
      *
-     * @param  RequestInterface $request
-     * @param  Table            $table
-     * @param  array            $documents
+     * @param int $storeId
      *
      * @return array
      */
-    private function getAggregations($request, $table, $documents)
+    private function getContextParams($storeId)
     {
-        if (version_compare($this->config->getMagentoVersion(), '2.1.0', '<') === true) {
-            return $this->aggregationBuilder->build($request, $table);
-        }
+        $contextParams = [];
+        $category = $this->registry->registry('current_category');
+        $page = !is_null($this->request->getParam('page')) ?
+            (int) $this->request->getParam('page') - 1 :
+            0;
+
+        if ($category) {
```

Comment:
Most of the context parameters here are set only on a category page (where `$category` is set), but facets and other parameters can be set on search results page as well.